### PR TITLE
Improving the behaviour of an empty InferenceData object. (Issue no 486).

### PR DESCRIPTION
**Pull request by Ban-zee**
Addressing issue #486

### CODE REVIEW

**Comment by canyon289** [comment]
@Ban-zee Please add tests as well

**Comment by Ban-zee** [comment]
Definitely, will keep that in mind in the future. I'll make the necessary changes.

**Comment by canyon289** [comment]
@Ban-zee If you want to practice you can change the title of this PR and amend your commit to have a more useful phrasing :)

**Comment by canyon289** [comment]
I would add a meaningful test for these lines that are currently not tested
https://coveralls.io/builds/21387295/source?filename=arviz%2Fdata%2Fio_netcdf.py#L47

Reading this over a bit more too can we add a doc string that briefly explains why the if else is there?

**Comment by canyon289** [comment]
@Ban-zee Thanks for the title change :)
If you need help with the other asks let us know! We're here to help

**Comment by Ban-zee** [comment]
Sorry for keeping this PR open for so long, I was a bit busy this past
week. Regarding the addition of tests, it would be helpful if anyone could
guide me through that process.

**Comment by canyon289** [comment]
@Ban-zee Don't worry about how long you keep the PR open, there's no timeline in open source!

For testing, are you familiar with pytest or testing principles in general? Trying to get a baseline of where you're at so I can guide you appropriately

**Comment by Ban-zee** [comment]
Thinking of pushing my commits with the other changes you suggested(other
than the tests), but git keeps rejecting it as the tip of my local branch
is behind the remote. Tried resolving it but nothing seem to work. Should I
just force push?
Regarding the tests, I have gone through the pytest documentation and have
looked into the tests included in this project. It would be helpful if you
could tell me how to add my tests to the ones which already exist (in my
case mainly for the else block).

**Comment by canyon289** [comment]
When you do a "destructive" git command like amend you have to force push. Just be careful because if you mess up the command its becomes more difficult to go back, and in some cases impossible. 

For the data functions I would look at this module and this test. It'll give you an idea how things are implemented

https://github.com/arviz-devs/arviz/blob/master/arviz/tests/test_data.py#L651

**Comment by ahartikainen** [comment]
Let's say it is impossible. (It is just really really hard, but doable)

**Comment by Ban-zee** [comment]
I did use amend. I'll have to force push anyway now.

**Comment by Ban-zee** [comment]
Getting the following error: "No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself." Any idea on what should be done?

**Comment by ahartikainen** [comment]
That happens sometimes (travis freezes). Restarting the job usually works (might need admin rights, so usually someone from the devteam can do it).

**Comment by Ban-zee** [comment]
Thank you.

**Comment by ColCarroll** [comment]
Hrm... fails on Python 3.5, but not Python 3.6. I restarted in case it is a flaky test, but here's the message:

```pythontb
=================================== FAILURES ===================================
______________________ TestIONetCDFUtils.test_io_function ______________________
self = <arviz.tests.test_data.TestIONetCDFUtils object at 0x7f4bfb122668>
data = <class 'arviz.tests.test_data.TestIONetCDFUtils.data.<locals>.Data'>
eight_schools_params = {'J': 8, 'sigma': array([15., 10., 16., 11.,  9., 11., 10., 18.]), 'y': array([28.,  8., -3.,  7., -1.,  1., 18., 12.])}
    def test_io_function(self, data, eight_schools_params):
        inference_data = self.get_inference_data(  # pylint: disable=W0612
            data, eight_schools_params
        )
        assert hasattr(inference_data, "posterior")
        here = os.path.dirname(os.path.abspath(__file__))
        data_directory = os.path.join(here, "saved_models")
        filepath = os.path.join(data_directory, "io_function_testfile.nc")
        # az -function
        to_netcdf(inference_data, filepath)
        assert os.path.exists(filepath)
        assert os.path.getsize(filepath) > 0
        inference_data2 = from_netcdf(filepath)
>       assert hasattr(inference_data2, "posterior")
E       AssertionError: assert False
E        +  where False = hasattr(Inference data with groups:\n	> posterior_predictive, 'posterior')
```

**Comment by Ban-zee** [comment]
Any idea on why the test fails in python 3.5 but not in 3.6?

File: arviz/data/inference_data.py [hunk 1]
```diff
@@ -80,15 +80,20 @@ def to_netcdf(self, filename, compress=True):
...
+        if self._groups:  # check's whether a group is present or not.
+            for group in self._groups:
+                data = getattr(self, group)
+                kwargs = {}
+                if compress:
+                    kwargs["encoding"] = {var_name: {"zlib": True} for var_name in data.variables}
+                data.to_netcdf(filename, mode=mode, group=group, **kwargs)
+                data.close()
+                mode = "a"
+                return filename
```
Comment (ahartikainen) [inline arviz/data/inference_data.py]: here is indentation error.

You could return `filename` at the end of the function.

File: arviz/data/inference_data.py [hunk 2]
```diff
@@ -80,15 +80,20 @@ def to_netcdf(self, filename, compress=True):
...
+        else:  # creates a netcdf file for an empty InferenceData object.
+            empty_netcdf_file = nc.Dataset(filename, mode="w", format="NETCDF4")
+            empty_netcdf_file.close()
+            return filename
```
Comment (ahartikainen) [inline arviz/data/inference_data.py]: remove the first `return filename` and return this outside the if-else block. (intend back)

**Comment by Ban-zee** [comment]
Made the changes. Thank you for your help.

**Comment by Ban-zee** [comment]
Any other change I have to make before merging?

**Comment by ColCarroll** [comment]
This looks good -- thank you for the contribution @Ban-zee!

**Pull request merged**